File: NeuroSonic/Startup/SplashScreen.cs
using theori;
using theori.Graphics;
using theori.IO;

using NeuroSonic.Platform;

namespace NeuroSonic.Startup
{
    sealed class SplashScreen : Layer
    {
        const float WAIT_TIME = 0.25f;
        const float TRANSITION_TIME = 0.5f;
        const float HOLD_TIME = 2.0f;
        const float TOTAL_TIME = WAIT_TIME + TRANSITION_TIME + HOLD_TIME;

        private BasicSpriteRenderer? m_renderer;

        private float m_timer = 0.0f, m_alpha = 0.0f;

        public SplashScreen()
        {
        }

        public override void Initialize()
        {
            m_renderer = new BasicSpriteRenderer();
        }

        public override bool KeyPressed(KeyInfo info)
        {
            if (info.KeyCode == KeyCode.ESCAPE)
                Host.Exit();
            else m_timer = TOTAL_TIME;

            return true;
        }

        public override void Update(float delta, float total)
        {
            m_timer += delta;
            if (m_timer > TOTAL_TIME)
            {
                ClientAs<NscClient>().CloseCurtain(() => Push(new NeuroSonicStandaloneStartup()));
                SetInvalidForResume();
            }
            else
            {
                float timer = m_timer;
                if (timer >= WAIT_TIME)
                {
                    timer -= WAIT_TIME;
                    if (timer >= TRANSITION_TIME)
                    {
                        timer -= TRANSITION_TIME;
                        if (timer >= HOLD_TIME)
                        {
                            // should have been triggered above
                        }
                        else m_alpha = 1;
                    }
                    else m_alpha = timer / TRANSITION_TIME;
                }
                else m_alpha = 0;
            }
        }

        public override void Render()
        {
            var r = m_renderer!;
            var text = theori.Host.StaticResources.GetTexture("textures/audfx-text-large");

            float width = Window.Width * 0.7f;
            float height = width * text.Height / text.Width;

            r.BeginFrame();
            r.SetImageColor(255, 255, 255, 255 * m_alpha);
            r.Image(text, (Window.Width - width) / 2, (Window.Height - height) / 2, width, height);
            r.EndFrame();
        }
    }
}

File: NeuroSonic/Resources/scripts/splashScreen.lua

local WAIT_TIME = 0.25;
local TRANSITION_TIME = 0.5;
local HOLD_TIME = 2.0;
local TOTAL_TIME = WAIT_TIME + TRANSITION_TIME + HOLD_TIME;

local alpha = 0;
local timer = 0;

local transitionFunction = function()
    if (theori.isFirstLaunch()) then
        theori.layer.push("initialSetup");
    else
        theori.layer.push("titleScreen");
    end
end;

function theori.layer.init()
    theori.input.controller.pressed:connect(function(controller, button)
        if (button == "start") then
            timer = TOTAL_TIME;
        end
    end);
end

function theori.layer.update(delta, total)
    timer = timer + delta;
    if (timer > TOTAL_TIME) then
        theori.layer.setInvalidForResume();
        theori.graphics.closeCurtain(0.25, transitionFunction);
    else
        local temp = timer;
        if (temp >= WAIT_TIME) then
            temp = temp - WAIT_TIME;
            if (temp >= TRANSITION_TIME) then
                temp = temp - TRANSITION_TIME;
                if (temp < HOLD_TIME) then
                    alpha = 1;
                end
            else
                alpha = temp / TRANSITION_TIME;
            end
        else
            alpha = 0;
        end
    end
end

function theori.layer.render()
    local text = theori.graphics.getStaticTexture("audfx-text-large");
    
    local w, h = theori.graphics.getViewportSize();

    local width = w * 0.7;
    local height = width * text.Height / text.Width;
    
    theori.graphics.setFillToTexture(text, 0, 169, 255, 255 * alpha);
    theori.graphics.fillRect((w - width) / 2, (h - height) / 2, width, height);
end